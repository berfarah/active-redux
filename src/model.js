import Store from './store';
import Registry from './registry';

/**
 * @private
 */
const invalidAttribute = (field) => (
  new Error(`${field} needs an attribute type`)
);

/**
 * @private
 */
const defineAttribute = (object, field, attribute) => {
  Object.defineProperty(object, field, {
    get() {
      return attribute.cast(this.data.attributes[field]);
    }
  });
};

/**
 * @private
 */
const defineRelationship = (object, field) => {
  Object.defineProperty(object, field, {
    get() {
      const { isArray, resource } = this.constructor.attributes[field];
      const model = Registry.get(resource);
      if (!model) {
        throw new Error(`Unregistered model: ${resource}`);
      }
      const data = this.data.relationships[field].data;

      return isArray
        ? Store.where({ id: data.map((d) => d.id) }, { model })
        : Store.find({ id: data.id }, { model });
    }
  });
};

/**
 * Define JSON-API attributes and relationships
 * @private
 */
const defineMethods = (object, attributes) => {
  const relationships = {};

  Object.entries(attributes).forEach(([field, attribute]) => {
    switch (attribute.type) {
      case 'attribute':
        defineAttribute(object.prototype, field, attribute);
        break;
      case 'relationship':
        defineRelationship(object.prototype, field);
        relationships[attribute.resource] = { key: field, ...attribute };
        break;
      default:
        throw invalidAttribute(field);
    }
  });

  return relationships;
};

/**
 * @private
 */
const copyClassName = (from, to) => {
  Object.defineProperty(to, 'name', {
    value: from.name,
    writable: false,
    enumerable: false,
    configurable: false,
  });
};

/**
 * Helper function to create endpoint URIs
 * @private
 * @example
 * const obj = { type: 'animal', id: 5 }
 * parseParams(obj, '/api/v1/:type/:id')
 * // => '/api/v1/animal/5'
 */
export const parseParams = (context, string) => {
  const regexp = /:[^/.]+/g;
  return string.replace(regexp, (match) => context[match.substring(1)]);
};

/**
  * Defines a model
  * @alias module:active-redux.define
  * @example
  * const Person = ActiveRedux.define('people', class Person {});
  * @param {string} type - JSON-API type for the model
  * @param {Class} [model] - Class to extend
  * @return {Model}
  */
export default function define(type, model = class {}) {
  /**
   * Active Redux Model
   */
  const Model = class extends model {

    /**
    * JSON-API type of the model
    * @private
    */
    static type = type;

    /**
    * Attributes of a model, for which methods will be defined
    * @example
    * import Model, { Attr } from 'active-redux';
    *
    * const Person = Model.define('people', class Person {
    *   static attributes = {
    *     name: Attr.string(),
    *   }
    * });
    *
    * const joe = new Person({ attributes: { name: "Joe" } });
    * joe.name // => "Joe"
    */
    static attributes = model.attributes || {};

    /**
    * Endpoints that will be queried. Can be overwritten.
    * @private
    */
    static endpoints = model.endpoints || {
      create: ':type',
      read: ':type',
      update: ':type/:id',
      delete: ':type/:id',
    };

    /**
    * @private
    */
    static relationships = {};

    /**
    * Gets all of that resource
    * @example
    * Person.all()
    * // => Promise<Array<Person>>
    * @return {Promise<Array<this>>} Array of model instances
    */
    static all() {
      return Store.all({ model: this });
    }

    /**
    * Queries the store for that resource
    * @example
    * Person.where({ name: "Joe" })
    * // => Promise<Array<Person>>
    * @param {Object} query - Query for the store
    * @param {Object} [options]
    * @param {boolean} [options.remote] - Call to API if no records found?
    * @return {Promise<Array<this>>} Array of model instances
    */
    static where(query = {}, options = {}) {
      return Store.where(query, { model: this, ...options });
    }

    /**
    * Gets one of that resource
    * @example
    * Person.find({ id: 5 })
    * // => Promise<Person>
    * @param {Object} query - Query for the store
    * @param {Object} [options]
    * @param {boolean} [options.remote] - Call to API if no records found?
    * @return {Promise<(this|null)>} A model instance
    */
    static find(query = {}, options = {}) {
      return Store.find(query, { model: this, ...options });
    }

    /**
    * Find the endpoint for a specific action
    * @private
    */
    static endpoint(action, params = this) {
      return parseParams(params, this.endpoints[action]);
    }

    /**
    * @private
    * @see Model.endpoint
    */
    endpoint(action) {
      return this.constructor.endpoint(action, this);
    }

    constructor(data) {
      super(data);
      this.data = data;
    }

    /**
     * @return {string} JSON-API Resource ID
     */
    get id() {
      return this.data.id;
    }

    /**
     * @return {string} JSON-API Resource Type
     */
    get type() {
      return this.constructor.type;
    }
  };

  copyClassName(model, Model);
  Model.relationships = defineMethods(Model, Model.attributes);
  Registry.register(Model);
  return Model;
}
