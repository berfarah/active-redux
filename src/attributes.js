import yup from 'yup';

export const string = ({ default: defaultValue } = {}) => (obj) => (
  yup.string().default(defaultValue).nullable().cast(obj)
);

export const number = ({ default: defaultValue } = {}) => (obj) => (
  yup.number().default(defaultValue).nullable().cast(obj)
);

export const date = ({ default: defaultValue } = {}) => (obj) => {
  if (obj === undefined) return defaultValue;
  if (obj === null) return obj;
  if (typeof obj === 'number') return new Date(obj);
  return yup.date().cast(obj)
};

export const boolean = ({ default: defaultValue } = {}) => (obj) => (
  typeof obj === 'string' ?
    Boolean(obj) :
    yup.boolean().default(defaultValue).nullable().cast(obj)
);

export const array = () => (obj) => (
  Array.isArray(obj) ? obj : [obj]
);

export const object = () => (obj) => (
  typeof obj === 'object' ? obj : {}
);

