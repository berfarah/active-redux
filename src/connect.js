import React from 'react';
import { connect } from 'react-redux';

/**
 * Map async actions to props
 * @alias module:active-redux.connect
 * @example
 * import { connect } from 'active-redux';
 * import Article from '../models/article';
 *
 * const List = ({ isLoading, articles }) => {
 *   if (isLoading) return (
 *     <div>Loading...</div>
 *   );
 *
 *   return (
 *     <ul>
 *       {articles.map((article) => (
 *         <li><a href={article.link}>{article.title}</li>
 *       ))}
 *     </ul>
 *   );
 * };
 *
 * export default connect({
 *   articles: (props) => Article.all()
 * })(List);
 * @param {Object.<String, Function>} promises
 * @param {Object} options
 * @param {Number} options.delay Delay in Milliseconds. The component will be empty during this
 *   time. This is in place because people perceive a waiting time as longer when there is
 *   immediately a loading screen
 * @param {Number} options.timeout Timeout in Milliseconds.
 * @return {Function} Function that takes a component
 */
export default (
  promises = {},
  {
    delay = 150,
    timeout,
  } = {},
) => (WrappedComponent) => connect(
  ({ api: { resources } }) => ({ resources })
)(
  class PromisedComponent extends React.Component {
    static WrappedComponent = WrappedComponent;
    static displayName = `Promised(${WrappedComponent.displayName || WrappedComponent.name})`;
    static propTypes = WrappedComponent.propTypes;

    constructor(props) {
      super(props);

      this.state = {
        isLoading: true,
        hasTimedOut: false,
        hasDelayed: false,
        responses: {},
        errored: {},
      };

      this.promiseCount = Object.keys(promises).length;
      this.loadedCount = 0;
    }

    componentWillMount() {
      if (!this.state.isLoading) return;
      this._isMounted = true;
      this._setupDelay();
      this._setupTimeout();
      this._startPromises(this.props);
    }

    componentWillUnmount() {
      this._isMounted = false;
      clearTimeout(this._delay);
      clearTimeout(this._timeout);
    }

    componentWillReceiveProps(nextProps) {
      if (this.props.resources !== nextProps.resources) {
        this._startPromises(nextProps);
      }
    }

    _startPromises(props) {
      const update = ({ key, data, error = false }) => {
        if (!this._isMounted || this.hasTimedOut) return;

        this.loadedCount += 1;
        this.setState({
          responses: {
            ...this.state.responses,
            [key]: data
          },
          errored: {
            ...this.state.errored,
            [key]: error
          },
          isLoading: this.promiseCount > this.loadedCount,
        });
      };

      Object.entries(promises)
        .forEach(([key, promise]) => promise(props)
          .then((data) => update({ key, data }))
          .catch((data) => update({ key, data, error: true }))
      );
    }

    _setupDelay() {
      if (delay) {
        this._delay = setTimeout(() => {
          this.setState({ hasDelayed: true });
        }, delay);
      }
    }

    _setupTimeout() {
      if (timeout) {
        this._timeout = setTimeout(() => {
          this.setState({ hasTimedOut: true });
        }, timeout);
      }
    }

    render() {
      const { responses, ...props } = this.state;
      const { resources, ...theseProps } = this.props;
      return (
        <WrappedComponent
          {...props}
          {...responses}
          {...theseProps}
        />
      );
    }
  }
);
