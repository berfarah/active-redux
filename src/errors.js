export const missingStore = () => (
  new Error('A store has not been registered in the Registry')
);

export const invalidAttribute = (field) => (
  new Error(`${field} needs an attribute type`)
);
