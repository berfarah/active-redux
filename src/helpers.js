export const deepPartialEqual = (obj1, obj2) =>
  Object.entries(obj1).reduce((equal, [key, value]) => {
    if (typeof value === 'object' && !Array.isArray(value)) {
      return deepPartialEqual(value, obj2[key]);
    }

    if (value !== obj2[key]) {
      return false;
    }

    return equal;
  }, true);
