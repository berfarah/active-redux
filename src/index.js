export Model from './model';
export Registry from './registry';
export * as Attr from './attributes';
export * from './api';
