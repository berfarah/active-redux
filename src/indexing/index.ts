export { default as extension } from './reducer';
export * from './reducer';
