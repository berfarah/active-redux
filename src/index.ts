import Store from './store';
import * as Attr from './attributes';
import * as apiTmp from './api';
import { extension as indexing } from './indexing';

export { default as define } from './model';

const api = Object.assign(
  {},
  apiTmp,
  { reducer: apiTmp.reducer.extend(indexing) },
);

/**
  * Registers the store to the model for .where/.find/.all queries
  * @function
  * @example
  * import { bind } from 'active-redux';
  * import store from './store';
  *
  * bind(store);
  * @param {Object} store - A redux store
  */
const bind = (store) => { Store.bind(store); };

export {
/**
 * @alias module:active-redux.Attr
 * @see module:active-redux/attributes
 */
  Attr,
  bind,
/**
 * @alias module:active-redux.api
 * @see module:active-redux/api
 */
  api,
};
