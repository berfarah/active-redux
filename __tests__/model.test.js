import { Attr, Model } from '../src';
import jsonApiData from '../fixtures/json-api-body';
import mockStore from '../fixtures/store';

const data = jsonApiData.data[0];
const [author, ...comments] = jsonApiData.included;

describe('Model', () => {
  let apiState = mockStore.getState().api;

  beforeEach(() => {
    Model.store = mockStore;
  });

  describe('.__defineMethods__()', () => {
    test('defines attribute methods', () => {
      class Article extends Model {
        static attributes = {
          title: Attr.string()
        };
      }
      expect(new Article(data).title).toEqual(undefined);
      Article.__defineMethods__();
      expect(new Article(data).title).toEqual(data.attributes.title);
    });

    test('defines hasOne methods', () => {
      class Article extends Model {
        static attributes = {
          author: Attr.hasOne(),
        };
      }

      expect(new Article(data).author).toBeUndefined();
      Article.__defineMethods__();
      const article = new Article(data);
      expect(article.author).toBeDefined();
      expect(typeof article.author).toEqual('object');
    });

      test('defines hasMany methods', () => {
        class Article extends Model {
          static attributes = {
            comments: Attr.hasMany(),
          }
        }

        expect(new Article(data).comments).toBeUndefined();
        Article.__defineMethods__();
        const article = new Article(data);
        expect(article.comments).toBeDefined();
        expect(article.comments).toHaveProperty('length');
      });

      test('should throw if attribute is invalid', () => {
        class Article extends Model {
          static attributes = {
            comments: {
              invalid: 'attribute',
            }
          }
        }

        expect(Article.__defineMethods__.bind(Article)).toThrow('comments needs an attribute type');
      });
  });

  describe('#type', () => {
    test('gets the static type', () => {
      class Article extends Model { static type = 'articles'; }
      expect(new Article().type).toEqual('articles');
    });
  });

  describe('#dispatch', () => {
    test('provides a getter for store dispatch', () => {
      expect(Model.dispatch).toBeInstanceOf(Function);
    });
  });

  describe('#findById', () => {
    test('should find an entity in store by id', () => {
      const response = Model.findById(5, 'comments');
      expect(response.body).toEqual(apiState.comments[5].body);
    });
  });

  describe('#findAll', () => {
    test('should return all values from a resource store', () => {
      const response = Model.findAll('comments');

      expect(response).toMatchSnapshot();
    });
  });

  describe('#find', () => {
    test('should return an array of values matching a query', () => {
      const response = Model.find({
        attributes: {
          body: 'First!',
        },
      }, 'comments');
      expect(response).toMatchSnapshot();
    });

    test('should allow arrays as lookup value', () => {
      const response = Model.find({
        id: ['5', '12'],
      }, 'comments');
      expect(response).toMatchSnapshot();
    });

    test('should search for nested properties', () => {
      const response = Model.find({
        attributes: {
          address: {
            road: '123 Main St'
          }
        }
      }, 'people');

      expect(response).toMatchSnapshot();
    });
  });
});
