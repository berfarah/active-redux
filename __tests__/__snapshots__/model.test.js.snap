// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model #find should allow arrays as lookup value 1`] = `
Array [
  Model {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Model {
    "data": Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "links": Object {
        "self": "http://example.com/comments/12",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
]
`;

exports[`Model #find should return an array of values matching a query 1`] = `
Array [
  Model {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
]
`;

exports[`Model #find should search for nested properties 1`] = `
Array [
  Model {
    "data": Object {
      "attributes": Object {
        "address": Object {
          "road": "123 Main St",
        },
        "first-name": "Dan",
        "last-name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "links": Object {
        "self": "http://example.com/people/9",
      },
      "type": "people",
    },
  },
]
`;

exports[`Model #findAll should return all values from a resource store 1`] = `
Array [
  Model {
    "data": Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "links": Object {
        "self": "http://example.com/comments/5",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
  Model {
    "data": Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "links": Object {
        "self": "http://example.com/comments/12",
      },
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  },
]
`;
